buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectGenerate.java

```diff
@@ -0,0 +1,403 @@
+    String fileText = new String(Files.readAllBytes(Paths.get(file.getAbsolutePath())),
+        StandardCharsets.UTF_8);
+    Queue<MatchResult> declarations = new LinkedList();
+    Matcher matcher = declarationPattern.matcher(fileText);
+    while (matcher.find()) {
+      declarations.add(matcher.toMatchResult());
+    }
+    Queue<String> tokens = new LinkedList(Arrays.asList(tokenizerPattern.split(fileText)));
+    int charIndex = 0;
+    while (!declarations.isEmpty()) {
+      MatchResult declaration = declarations.poll();
+      while (charIndex < declaration.start()) {
+        charIndex += tokens.poll().length();
+      }
+      charIndex = processFunction(tokens, functionMap, charIndex,
+          declaration.end() - 1,  getFunctionName(declaration.group()));
+    }
+  }
+
+  /**
+   * Parses a function of the form:
+   *
+   * <return_type> <name>(<args>) :
+   * {
+   *     <properties>
+   * }
+   * {
+   *     <body>
+   * }
+   *
+   * @param tokens The tokens starting from the function declaration and
+   *               ending at EOF
+   * @param functionMap The map to which the extracted functions will be added to
+   * @param charIndex The character index of the entire text of the file at which
+   *                  the parsing is commencing at
+   * @param declarationEnd The char index (of entire file text) at which the
+   *                       function declaration ends
+   * @param functionName The name of the function being processed
+   */
+  private int processFunction(
+      Queue<String> tokens,
+      Map<String, String> functionMap,
+      int charIndex,
+      int declarationEnd,
+      String functionName) {
+    StringBuilder functionBuilder = new StringBuilder();
+    // Process the declaration:
+    while (charIndex < declarationEnd && !tokens.isEmpty()) {
+      String token = tokens.poll();
+      functionBuilder.append(token);
+      charIndex += token.length();
+    }
+    // Called twice as there are two curly blocks, one for initialization
+    // and one for the body.
+    charIndex = processCurlyBlock(functionBuilder, tokens, charIndex);
+    charIndex = processCurlyBlock(functionBuilder, tokens, charIndex);
+
+    functionMap.put(functionName, functionBuilder.toString());
+    return charIndex;
+  }
+
+  /**
+   * Parses a block of text surrounded by curly braces. Creates an instance of
+   * CurlyParser() that is used keep track of whether or not the curly block
+   * has been fully parsed or not. The current character index of the file text
+   * is updated and returned once parsing is complete.
+   *
+   * @param functionBuilder The builder unto which the tokens get added to once parsed
+   * @param tokens The tokens starting from the function declaration and
+   *               ending at EOF
+   * @param charIndex The character index of the entire text of the file at which
+   *                  the parsing is commencing at
+   */
+  private int processCurlyBlock(
```

Comment:
Should we update this function signature to indicate that it throws an IllegalStateException?